### ISSUE TITLE
tensor.repeat with zero repeats causes floating-point exception

### ISSUE THREAD

**Issue by cooijmanstim** [Theano/Theano#4149]
This triggers a floating-point exception on both CPU and GPU:

``` python
import theano, theano.tensor as T, numpy as np
i = theano.shared(0, name="i")
y = T.repeat(T.constant([1]), i, axis=0)
z, _ = theano.scan(lambda y: y, sequences=[y])
f = theano.function([], z)
print f()
```

**Comment by sygi** [comment]
I tried to take a look at it and found out that the problem exists only if one compiles with optimizer (setting 
`optimizer=fast_compile` gives `NotImplementedError: We didn't implemented yet the case where scan do 0 iteration` or `No code run for for{cpu,scan_fn}(Subtensor{int64}.0, Subtensor{:int64:}.0, Subtensor{int64}.0)` if run in debugMode (what is strange by itself).

It seems that some optimization inserted this node to the graph:
`Elemwise{Composite{Switch(EQ(i0, i1), (i0 // (-i0)), i0)}} [id A] <TensorType(int64, scalar)> ''`, which leads to `0 // 0`.

I tried to find out, which one it was, but commenting out pieces of `theano/tensor/opt.py` until the graph didn't contain that node was not an effective way. Could you suggest me how to look for the nasty optimization?

**Comment by nouiz** [comment]
Good investigation. But I think we should let the optimization do as it do now, but convert the error to Nan? Do you understand why this generate an error instead of setting a nan?

To see the list of optimization applied, you can use this Theano flag: optimizer_verbose=True. To disable some optiization, you can use this flag: optimizer_excluding=opt1:opt2

**Comment by lamblin** [comment]
`//` is the integer division, and the type is `int64`, so `NaN` is not a possible value here.
That part of the graph is probably inserted by code trying to compute the length of scan, or of a subtensor of scan. That code is probably in `theano/scan_module/scan_opt.py`.

**Comment by sygi** [comment]
Thanks for the hints. I played with turning different optimizations on and off and as far as I see, the problem arises in connection with shape inference.
Adding `local_shape_to_shape_i` to fast_compile gives me the crash:
```THEANO_FLAGS="optimizer=fast_compile,exception_verbosity=high,optimizer_including=local_shape_to_shape_i,optimizer_excluding=constant_folding:MergeOptimizer:local_subtensor_make_vector:local_useless_elemwise:constant_folding,optimizer_verbose=True" nosetests test_extra_ops:TestRepeatOp.test_zero_repeat -s```

The generated graph looks then like that:
```
local_shape_to_shape_i Shape.0 MakeVector{dtype='int64'}.0
local_shape_to_shape_i Shape.0 MakeVector{dtype='int64'}.0
for{cpu,scan_fn} [id A] ''   20
 |Subtensor{int64} [id B] ''   17
 | |MakeVector{dtype='int64'} [id C] ''   16
 | | |Elemwise{sub,no_inplace} [id D] ''   15
 | |   |Elemwise{switch,no_inplace} [id E] ''   14
 | |   | |Elemwise{eq,no_inplace} [id F] ''   10
 | |   | | |Subtensor{int64} [id G] ''   7
 | |   | | | |MakeVector{dtype='int64'} [id H] ''   6
 | |   | | | | |Elemwise{mul,no_inplace} [id I] ''   4
 | |   | | | |   |Subtensor{int64} [id J] ''   3
 | |   | | | |   | |MakeVector{dtype='int64'} [id K] ''   0
 | |   | | | |   | | |TensorConstant{1} [id L]
 | |   | | | |   | |Constant{0} [id M]
 | |   | | | |   |i [id N]
 | |   | | | |Constant{0} [id O]
 | |   | | |TensorConstant{-1} [id P]
 | |   | |Elemwise{int_div,no_inplace} [id Q] ''   13
 | |   | | |Elemwise{mul,no_inplace} [id R] ''   1
 | |   | | | |TensorConstant{1} [id L]
 | |   | | | |i [id N]
 | |   | | |Elemwise{neg,no_inplace} [id S] ''   12
 | |   | |   |Elemwise{mul,no_inplace} [id T] ''   9
 | |   | |     |Subtensor{int64} [id G] ''   7
 | |   | |Subtensor{int64} [id G] ''   7
 | |   |TensorConstant{0} [id U]
 | |Constant{0} [id V]
 |Subtensor{:int64:} [id W] ''   19
 | |Subtensor{int64::} [id X] ''   11
 | | |Reshape{1} [id Y] ''   8
 | | | |Alloc [id Z] ''   5
 | | | | |DimShuffle{0,x} [id BA] ''   2
 | | | | | |TensorConstant{(1,) of 1} [id BB]
 | | | | |Subtensor{int64} [id J] ''   3
 | | | | |i [id N]
 | | | |MakeVector{dtype='int64'} [id H] ''   6
 | | |Constant{0} [id BC]
 | |ScalarFromTensor [id BD] ''   18
 |   |Subtensor{int64} [id B] ''   17
 |Subtensor{int64} [id B] ''   17
```


I also tried to manually specify shape for the `i` variable in the test above, but then I got violated assumptions in [SpecifyShape op's make_node](https://github.com/Theano/Theano/blob/master/theano/compile/ops.py#L810)

**Comment by lamblin** [comment]
It seems to be fixed now, maybe by #5854.
Now `NotImplementedError: We didn't implemented yet the case where scan do 0 iteration`.
